### PR TITLE
Add simple ensures, requires, safety predicates

### PR DESCRIPTION
**Pull request by celinval**
I also added one proof-of-concept harness to ensure Kani can verify it.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/contracts/safety/build.rs [hunk 1]
```diff
@@ -0,0 +1,4 @@
+fn main() {
+    // We add the configurations here to be checked.
+    println!("cargo:rustc-check-cfg=cfg(kani_host)");
+}
```
Comment (feliperodri) [inline library/contracts/safety/build.rs:4]:
```suggestion
}
```

File: library/contracts/safety/src/kani.rs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+use quote::{quote, format_ident};
+use syn::{ItemFn, parse_macro_input};
+
+pub(crate) fn requires(attr: TokenStream, item: TokenStream) -> TokenStream {
+    rewrite_attr(attr, item, "requires")
+}
+
+pub(crate) fn ensures(attr: TokenStream, item: TokenStream) -> TokenStream {
+    rewrite_attr(attr, item, "ensures")
+}
```
Comment (feliperodri) [inline library/contracts/safety/src/kani.rs:11]: We won't include `modifies`?

File: library/contracts/safety/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+#[proc_macro_attribute]
+pub fn requires(attr: TokenStream, item: TokenStream) -> TokenStream {
+    tool::requires(attr, item)
+}
+
+#[proc_macro_error]
+#[proc_macro_attribute]
+pub fn ensures(attr: TokenStream, item: TokenStream) -> TokenStream {
+    tool::ensures(attr, item)
+}
```
Comment (feliperodri) [inline library/contracts/safety/src/lib.rs:25]: No `modifies`?

File: library/contracts/safety/src/runtime.rs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+/// For now, runtime requires is a no-op.
+///
+/// TODO: At runtime the `requires` should become an assert unsafe precondition.
+pub(crate) fn requires(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    item
+}
+
+/// For now, runtime requires is a no-op.
+///
+/// TODO: At runtime the `ensures` should become an assert as well.
```
Comment (feliperodri) [inline library/contracts/safety/src/runtime.rs:12]: I'd suggest to create a GitHub issue to track these TODOs.

File: library/core/src/ptr/mod.rs [hunk 5]
```diff
@@ -1766,6 +1769,8 @@ pub unsafe fn read_volatile<T>(src: *const T) -> T {
 #[stable(feature = "volatile", since = "1.9.0")]
 #[rustc_diagnostic_item = "ptr_write_volatile"]
 #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+// What happens to the value that was replaced? No drop?
```
Comment (feliperodri) [inline library/core/src/ptr/mod.rs:1772]: What does that mean?

File: library/core/src/ptr/mod.rs [hunk 6]
```diff
@@ -2290,3 +2295,20 @@ pub macro addr_of($place:expr) {
...
+#[cfg(kani)]
+#[unstable(feature="kani", issue="none")]
+mod verify {
+    use crate::fmt::Debug;
+    use super::*;
+    use crate::kani;
+
+    #[kani::proof_for_contract(read_volatile)]
+    pub fn check_read_u128() {
+        let val = kani::any::<u16>();
```
Comment (feliperodri) [inline library/core/src/ptr/mod.rs:2308]: I imagine this is just a place holder, so we don't want to explore other types.

File: library/core/src/ub_checks.rs [hunk 7]
```diff
@@ -160,3 +160,49 @@ pub(crate) const fn is_nonoverlapping(
...
+/// Provide a few predicates to be used in safety contracts.
+///
+/// At runtime, they are no-op, and always return true.
+#[cfg(not(kani))]
+mod predicates {
+    /// Check if a pointer can be dereferenced. I.e.:
+    ///   * `src` must be valid for reads. See [crate::ptr] documentation.
+    ///   * `src` must be properly aligned. Use read_unaligned if this is not the case.
+    ///   * `src` must point to a properly initialized value of type T.
+    ///
```
Comment (feliperodri) [inline library/core/src/ub_checks.rs:175]:
```suggestion
    /// Checks if a pointer can be dereferenced, ensuring:
    ///   * `src` is valid for reads (see [`crate::ptr`] documentation).
    ///   * `src` is properly aligned (use `read_unaligned` if not).
    ///   * `src` points to a properly initialized value of type `T`.
    ///
    /// [`crate::ptr`]: https://doc.rust-lang.org/std/ptr/index.html
```

File: library/core/src/ub_checks.rs [hunk 8]
```diff
@@ -160,3 +160,49 @@ pub(crate) const fn is_nonoverlapping(
...
+    pub fn can_dereference<T>(src: *const T) -> bool {
+        let _ = src;
+        true
+    }
+
+    /// Check if a pointer can be written to:
+    /// * `dst` must be valid for writes.
+    /// * `dst` must be properly aligned. Use [`write_unaligned`] if this is not the
```
Comment (feliperodri) [inline library/core/src/ub_checks.rs:183]:
```suggestion
    /// * `dst` must be properly aligned. Use `write_unaligned` if this is not the
```

File: library/core/src/ub_checks.rs [hunk 9]
```diff
@@ -160,3 +160,49 @@ pub(crate) const fn is_nonoverlapping(
...
+mod predicates {
+    /// Check if a pointer can be dereferenced. I.e.:
+    ///   * `src` must be valid for reads. See [crate::ptr] documentation.
+    ///   * `src` must be properly aligned. Use read_unaligned if this is not the case.
+    ///   * `src` must point to a properly initialized value of type T.
+    ///
+    pub fn can_dereference<T>(src: *const T) -> bool {
+        let _ = src;
+        true
+    }
```
Comment (feliperodri) [inline library/core/src/ub_checks.rs:179]: Why not the following pattern for all functions?
```suggestion
    pub fn can_dereference<T>(_src: *const T) -> bool {
        true
    }
```

File: library/core/src/ub_checks.rs [hunk 10]
```diff
@@ -160,3 +160,49 @@ pub(crate) const fn is_nonoverlapping(
...
+
+    /// Check if a pointer can be the target of unaligned reads.
+    /// * `src` must be valid for reads.
+    /// * `src` must point to a properly initialized value of type `T`.
+    pub fn can_read_unaligned<T>(src: *const T) -> bool {
+        let _ = src;
+        true
+    }
+
+    /// * `dst` must be valid for writes.
```
Comment (feliperodri) [inline library/core/src/ub_checks.rs:198]:
```suggestion
    /// Check if a pointer can be the target of unaligned writes.
    /// * `dst` must be valid for writes.
```

**Review by feliperodri — APPROVED** [review]

File: library/contracts/safety/src/kani.rs (hunk 2 above)
Comment (celinval) [reply library/contracts/safety/src/kani.rs:11]: Not for now... in those cases, we will have to add the Kani specific modifies. Is that correct @pnkfelix?

File: library/contracts/safety/src/lib.rs (hunk 3 above)
Comment (celinval) [reply library/contracts/safety/src/lib.rs:25]: see my answer above

File: library/core/src/ptr/mod.rs (hunk 5 above)
Comment (celinval) [reply library/core/src/ptr/mod.rs:1772]: Oops... that was a previous note. Let me delete it.

File: library/core/src/ptr/mod.rs (hunk 6 above)
Comment (celinval) [reply library/core/src/ptr/mod.rs:2308]: yes, this was just a proof of concept.

File: library/core/src/ub_checks.rs (hunk 9 above)
Comment (celinval) [reply library/core/src/ub_checks.rs:179]: so the documentation is nicer. :)

File: library/core/src/ptr/mod.rs (hunk 5 above)
Comment (celinval) [reply library/core/src/ptr/mod.rs:1772]:
```suggestion
```

File: library/core/src/ub_checks.rs (hunk 9 above)
Comment (feliperodri) [reply library/core/src/ub_checks.rs:180]: I'm fine either way 😉

File: library/contracts/safety/src/runtime.rs (hunk 4 above)
Comment (celinval) [reply library/contracts/safety/src/runtime.rs:12]: Sure... I'll create one

Comment (celinval) [reply library/contracts/safety/src/runtime.rs:12]: https://github.com/model-checking/verify-rust-std/issues/17

**Pull request merged**